q2_feature_classifier/classifier.py

```diff
@@ -172,11 +175,33 @@ def _autodetect_orientation(reads, classifier, n=100,
     return chain(reversed_n_reads, (r.reverse_complement() for r in reads))
 
 
-def classify_sklearn(reads: DNAIterator, classifier: Pipeline,
-                     reads_per_batch: int=262144, n_jobs: int=1,
+def _autotune_reads_per_batch(reads, n_jobs):
+    # detect effective jobs. Will raise error if n_jobs == 0
+    n_jobs = sklearn.externals.joblib.effective_n_jobs(n_jobs)
+    # we really only want to calculate this if running in parallel
+    if n_jobs != 1:
+        seq_count = subprocess.run(
+            ['grep', '-c', '^>', str(reads)], check=True,
+            stdout=subprocess.PIPE)
+        return int(int(seq_count.stdout.decode('utf-8')) / n_jobs)
+    # otherwise reads_per_batch = large value (similar to original default)
+    else:
+        return 500000
+
+
+def classify_sklearn(reads: DNAFASTAFormat, classifier: Pipeline,
+                     reads_per_batch: int=0, n_jobs: int=1,
                      pre_dispatch: str='2*n_jobs', confidence: float=0.7,
                      read_orientation: str=None
                      ) -> pd.DataFrame:
+    # autotune reads per batch
+    if reads_per_batch == 0:
+        reads_per_batch = _autotune_reads_per_batch(reads, n_jobs)
+
+    # transform reads to DNAIterator
```

Comment:
Is this still needed here, since you're calling `grep` to count the reads?

Reply:
Ben originally set up `classify_sklearn` to accept a `DNAIterator` as input — so I added this step to avoid disturbing anything downstream. This seemed more straightforward and took less code than creating a copy of `reads` and transforming into `DNAFASTAFormat`. Am I missing something, though? e.g., can I get a path view from a `DNAIterator`?

Reply:
Sorry, I didn't catch that the view type had changed in this PR. This is good as-is.